portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -28,13 +35,61 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> messages = new ArrayList<String>();
-    messages.add("first message");
-    messages.add("second message");
-    messages.add("third message");
-    Gson gson = new Gson();
-    String jsonMessage = gson.toJson(messages);
+    ArrayList<Message> messages = new ArrayList<Message>();
+
+    Query query = new Query("Message");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("email");
+      String name = (String) entity.getProperty("name");
+      String message = (String) entity.getProperty("message");
+
+      Message messageObj = new Message(name, email, message);
+      messages.add(messageObj);
+    }
+
+    String jsonMessage = convertToJson(messages);
     response.setContentType("application/json;");
     response.getWriter().println(jsonMessage);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String email = getParameter(request, "email");
+    String name = getParameter(request, "name");
+    String message = getParameter(request, "message");
+
+    if (email.isEmpty() || name.isEmpty() || message.isEmpty()) {
+      response.setContentType("text/html;");
+      response.getWriter().println("<meta http-equiv='refresh' content='3;URL=index.html'>");
+      response.getWriter().println("<p style='color:red;'>All form fields must not be empty</p>");
+      return;
+    }
+
+    Entity newEntity = new Entity("Message");
+    newEntity.setProperty("email", email);
+    newEntity.setProperty("name", name);
+    newEntity.setProperty("message", message);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(newEntity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  private String getParameter(HttpServletRequest request, String name) {
```

Comment:
This function should probably have a more informative name - getParameterWithDefault?

Reply:
but I dont provide any custom on default value

Reply:
getParameterWithEmptyStringDefault? it should at least differentiate slightly from getParameter to hint at what extra things this function does